lang/src/org/partiql/lang/eval/visitors/StaticTypeInferenceVisitorTransform.kt

```diff
@@ -524,6 +525,93 @@ internal class StaticTypeInferenceVisitorTransform(
             return nAry.withStaticType(computeReturnTypeForFunctionCall(signature, functionArguments, nAry.metas))
         }
 
+        // Call agg : "count", "avg", "max", "min", "sum"
+        override fun transformExprCallAgg(node: PartiqlAst.Expr.CallAgg): PartiqlAst.Expr {
+            val nAry = super.transformExprCallAgg(node) as PartiqlAst.Expr.CallAgg
+            val funcName = nAry.funcName
+            // unwrap the type if this is a collectionType
+            val argType = when (val type = nAry.arg.getStaticType()) {
+                is CollectionType -> type.elementType
+                else -> type
+            }
+
+            return nAry.withStaticType(computeReturnTypeForAggFunc(funcName, argType))
+        }
+
+        private fun computeReturnTypeForAggFunc(aggFunc: SymbolPrimitive, argType: StaticType): StaticType {
+            // nested type is not supported by agg functions.
+            // i.e. sum(1,2,[3,4]) is not supported
+            val argTypeCheck: (StaticType) -> Boolean = {
+                it.isSubTypeOf(StaticType.unionOf(StaticType.MISSING, StaticType.NULL, StaticType.NUMERIC))
+            }
+            return when (aggFunc.text) {
+                // current implementation of count will always return a long
+                "count" -> StaticType.INT8
+                // max/min supports all type and the result depends on comparison.
+                "max", "min" -> argType
+                // current implementation of avg always return a decimal.
+                "avg" -> {
+                    if (argTypeCheck(argType)) StaticType.DECIMAL else error("Aggregate function AVG has incompatible input type")
+                }
+                "sum" -> {
```

Comment:
For computing `SUM`'s return type, could we reuse the `computeBinaryArithmeticResultType` function? Or is the logic slightly different?

Reply:
The logic is slightly different in how aggregate functions handles `MISSING` and `NULL`.